File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/String/67_AddBinary.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


c.. Solution o..
    ___ addBinary  a, b
        """ Recursively binary add.
        """
        __ l..(a) __ 0:
            r_ b
        __ l..(b) __ 0:
            r_ a

        __ a[-1] __ '1' and b[-1] __ '1':
            r_ self.addBinary(self.addBinary(a[:-1], b[:-1]), '1') + '0'
        ____ a[-1] __ '0' and b[-1] __ '0':
            r_ self.addBinary(a[:-1], b[:-1]) + '0'
        ____
            r_ self.addBinary(a[:-1], b[:-1]) + '1'


c.. Solution_2 o..
    ___ addBinary  a, b
        """Iteratively way.
        """
        carry_in, index = '0', 0
        result = ""

        _____ index < m..(l..(a), l..(b)) or carry_in __ '1':
            num_a = a[-1 - index] __ index < l..(a) else '0'
            num_b = b[-1 - index] __ index < l..(b) else '0'

            val = int(num_a) + int(num_b) + int(carry_in)
            result = str(val % 2) + result
            carry_in = '1' __ val > 1 else '0'
            index += 1
        r_ result


c.. Solution_3 o..
    ___ addBinary  a, b
        r_ bin(eval("0b" + a) + eval("0b" + b))[2:]


"""
"0"
"0"
"111000"
"111111111"
"""

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/String/67_AddBinary.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-06-30 16:15:49
 */

c.. Solution {
public:
    string addBinary(string a, string b) {
        string result = "";
        int c = 0, i = a.size() - 1, j = b.size() - 1;
        _____(i>=0 || j>=0 || c__1){
            c += i >= 0 ? a[i --] - '0' : 0;
            c += j >= 0 ? b[j --] - '0' : 0;
            result = char(c % 2 + '0') + result;
            c /= 2;
        }
        r_ result;
    }
};


c.. Solution_2 {
public:
    string addBinary(string a, string b) {
        __(a.size()__0){
            r_ b;
        }
        __(b.size()__0){
            r_ a;
        }
        __(a[a.size()-1] __ '1' && b[b.size()-1] __ '1'){
            r_ addBinary(addBinary(
                string(a.begin(), a.end()-1), string(b.begin(), b.end()-1)), "1") + "0";
        }
        __(a[a.size()-1] __ '0' && b[b.size()-1] __ '0'){
            r_ addBinary(string(a.begin(), a.end()-1), string(b.begin(), b.end()-1)) + "0";
        }
        else{
            r_ addBinary(string(a.begin(), a.end()-1), string(b.begin(), b.end()-1)) + "1";
        }
    }
};

/*
"0"
"0"
"111000"
"111111111"
*/
